blog/features/replica-signed-queries.md

```diff
@@ -0,0 +1,149 @@
+
+## How do replica signed queries work?
+
+First, a list of node IDs and their corresponding public keys is put into the subnet's state tree so that they can be certified by the subnet.
+
+Users can obtain the certificate to validate the node's keys through an HTTP `read_state` call that includes a timestamp value. 
+
+To support replica signed queries, the query response format has been changed. Previously, query responses used the following format:
+
+```
+pub enum HttpQueryResponse {
+    Replied {
+        reply: HttpQueryResponseReply,
+    },
+    Rejected {
+        error_code: String,
+        reject_code: u64,
+        reject_message: String,
+    },
+}
+```
+
+The actual portion of the call that is signed is the HttpQueryResponse from the execution of the call. If the call is replied to, the `reply` field is included. If the call is rejected, the `error_code`, `reject_code`, and `reject_message` fields will be included in the response. 
+
+Queries also include a request ID, which is a SHA256 hash of the query's contents. With this `request_id` value, signatures for one query cannot be confused or exchanged for another.  Query responses will also include a timestamp value. 
+
+Query responses will now use the following format:
+
+```
+struct  NodeSignature {
+    // the batch time
+    timestamp: Time,
+    // the actual signature
+    signature: Blob,
+    // the ID of the node that created the signature
+    identity: NodeId
+    }
+pub enum HttpQueryResponse {
+Replied {
+    reply: HttpQueryResponseReply,
+    signatures: Vec<NodeSignature>,
+},
+Rejected {
+    error_code: String,
+    reject_code: u64,
+    signatures: Vec<NodeSignature>,
+},
+}
+```
+
+Query response verification is one via agents, such as agent-rs and agent-js, and other agents developed by members of the IC community. Agents now introduce a new `AgentError` type that indicates a verification failure. 
+
+In agent-rs, query response is defined in the file `ic-agent/src/agent/replica_api.rs` as follows:
+
+```
+pub struct CallReply {
+   #[serde(with = "serde_bytes")]
+   pub arg: Vec<u8>,
+   pub signatures: Vec<NodeSignature>,
+}
+#[derive(Debug, Clone, Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
+pub struct RejectResponse {
+    /// The [reject code] returned by the replica.
+    pub reject_code: RejectCode,
+    /// The rejection message.
+    pub reject_message: String,
+    /// The optional [error code]returned by the replica.
+    #[serde(default)]
+    pub error_code: Option<String>,
+    pub signatures: Vec<NodeSignature>,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(tag = "status")]
+pub enum QueryResponse {
+   #[serde(rename = "replied")]
+   Replied { reply: CallReply },
+   #[serde(rename = "rejected")]
+   Rejected(RejectResponse),
+}
+```
```

Comment:
With the interface specification quoted above, there is no need to mention them separately in agent-rs and agent-js again because they are just the implementation of the same protocol in different languages.